pkg/webhook/admission/request_handler.go

```diff
@@ -0,0 +1,215 @@
+	cf := serializer.NewCodecFactory(scheme)
+	return &RequestHandler{
+		codecFactory: cf,
+		serializer:   apijson.NewSerializerWithOptions(apijson.DefaultMetaFactory, scheme, scheme, apijson.SerializerOptions{}),
+		decoder:      cf.UniversalDecoder(),
+		validator:    validator,
+		mutator:      mutator,
+	}
+}
+
+var _ handlers.ValidatingAdmissionHook = &RequestHandler{}
+var _ handlers.MutatingAdmissionHook = &RequestHandler{}
+
+// Validate will decode the Object (and OldObject, if set) in the AdmissionRequest into the
+// internal API version.
+// It will then invoke the validation handler to build a list of warning messages and any
+// errors generated during the admission chain.
+func (rh *RequestHandler) Validate(ctx context.Context, admissionSpec *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse {
+	status := &admissionv1.AdmissionResponse{}
+	status.UID = admissionSpec.UID
+	// short-path if there is no validator actually registered or the handler does not handle this operation.
+	if rh.validator == nil || !rh.validator.Handles(admissionSpec.Operation) {
+		status.Allowed = true
+		return status
+	}
+
+	// decode new version of object
+	obj, _, err := rh.decoder.Decode(admissionSpec.Object.Raw, nil, nil)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	// attempt to decode old object
+	var oldObj runtime.Object
+	if len(admissionSpec.OldObject.Raw) > 0 {
+		oldObj, _, err = rh.decoder.Decode(admissionSpec.OldObject.Raw, nil, nil)
+		if err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+				Message: err.Error(),
+			}
+			return status
+		}
+	}
+
+	warnings, err := rh.validator.Validate(ctx, *admissionSpec, oldObj, obj)
+	status.Warnings = warnings
+
+	// return with allowed = false if any errors occurred
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusNotAcceptable, Reason: metav1.StatusReasonNotAcceptable,
+			Message: err.Error(),
+		}
+		return status
+	}
+	status.Allowed = true
+	return status
+}
+
+func (rh *RequestHandler) Mutate(ctx context.Context, admissionSpec *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse {
+	status := &admissionv1.AdmissionResponse{}
+	status.UID = admissionSpec.UID
+	status.Allowed = true
+	// short-path if there is no mutator actually registered
+	// we still continue if the mutator does not handle the resource so scheme-registered
+	// defaulting functions are still run against the object.
+	if rh.mutator == nil {
+		status.Allowed = true
+		return status
+	}
+
+	obj, _, err := rh.decoder.Decode(admissionSpec.Object.Raw, nil, nil)
```

Comment:
This is where scheme-registered defaulting functions are actually run. The decoder here will decode to the *internal version*, as well as applying defaults to the resource in the API version that the request was originally in.

Reply:
Come to think.. a slight subtlety here that may be missed:

This will actually decode from the raw object bytes passed to the webhook. HOWEVER, this *may not necessarily* be the same version as denoted in `RequestResource` as the apiserver can perform conversion of a resource before submitting to mutation webhooks.

To avoid this, we'll need to:

1) decode into internal version without applying defaults
2) convert into `RequestResource` version
3) apply defaults in this API version
4) convert back into the internal version and execute mutating admission plugins

This doesn't effect *us* today anymore as we no longer have any versions other than v1, but this is a difficult to spot issue... the alternative workaround is to specify `versions: ["*"]` in our MutatingWebhookConfiguration object so that the apiserver does not call the conversion webhook before submitting to mutating webhooks (and instead just passes through the request as-is)

Reply:
Been digging into this some more and as I write a unit test to cover this case, it seems like it's actually really nasty (also tries to send a patch to overwrite the `apiVersion` and effectively generates a patch for a conversion during defaulting 😅).

I'm going to fix this now, but holding to prevent this bug merging :)

/hold

Reply:
Pushed up a unit test & a fix for this bug :)

/unhold